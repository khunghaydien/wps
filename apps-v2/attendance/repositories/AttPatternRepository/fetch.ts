import Api from '../../../commons/api';

import adapter from '@apps/repositories/adapters';

import {
  createFromRemote,
  IAttPatternRepository,
} from '@attendance/domain/models/AttPattern';

export type AttPattern = {
  name: string;
  code: string;
  startTime: number | null;
  endTime: number | null;
  rest1StartTime: number | null;
  rest1EndTime: number | null;
  rest2StartTime: number | null;
  rest2EndTime: number | null;
  rest3StartTime: number | null;
  rest3EndTime: number | null;
  rest4StartTime: number | null;
  rest4EndTime: number | null;
  rest5StartTime: number | null;
  rest5EndTime: number | null;
  workSystem: string;
  flexStartTime: number | null;
  flexEndTime: number | null;
  withoutCoreTime: boolean;
};

export type Response = Readonly<{
  requestableDayType: string;
  canDirectInputTimeRequest: boolean;
  patterns: AttPattern[];
}>;

export default (async ({
  targetDate,
  ignoredId,
  employeeId: empId,
}: {
  targetDate: string;
  ignoredId?: string;
  employeeId?: string;
}) => {
  const response: Response = await Api.invoke({
    path: '/att/daily-pattern/list',
    param: adapter.toRemote({
      targetDate,
      ignoredId: ignoredId || '',
      empId: empId || '',
    }),
  });
  const DailyAttPattern = {
    requestableDayType: response.requestableDayType,
    canDirectInputTimeRequest: response.canDirectInputTimeRequest,
    patterns: response.patterns.map((r) =>
      adapter.fromRemote({ ...r }, [createFromRemote])
    ),
  };
  return DailyAttPattern;
}) as IAttPatternRepository['fetch'];
