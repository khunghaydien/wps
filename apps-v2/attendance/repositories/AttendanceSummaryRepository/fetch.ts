import Api from '@apps/commons/api';

import {
  BaseAttendanceSummary,
  convert,
} from '../models/BaseAttendanceSummary';
import {
  AttendanceSummary as DomainAttendanceSummary,
  IAttendanceSummaryRepository,
  Status,
} from '@attendance/domain/models/AttendanceSummary';

export type Response = BaseAttendanceSummary & {
  summaryName: string;
  status: Status;
  hasCalculatedAbsence: boolean;
};

const $convert = ({
  hasCalculatedAbsence,
  status,
  summaryName,
  ...response
}: Response): DomainAttendanceSummary => ({
  ...convert(response),
  name: summaryName,
  status,
  hasCalculatedAbsence,
});

const fetch: IAttendanceSummaryRepository['fetch'] = async (param) => {
  const response: Response = await Api.invoke({
    path: '/att/summary/get',
    param: {
      empId: param?.employeeId,
      targetDate: param?.targetDate,
    },
  });
  return $convert(response);
};

export default fetch;
