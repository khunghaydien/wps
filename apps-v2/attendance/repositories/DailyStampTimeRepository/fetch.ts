import Api from '@apps/commons/api';

import * as CommuteCount from '@attendance/repositories/models/CommuteCount';

import {
  ClockType,
  DailyStampTime as DomainDailyStampTime,
  IDailyStampTimeRepository,
} from '@attendance/domain/models/DailyStampTime';
import * as DomainFixDailyRequest from '@attendance/domain/models/FixDailyRequest';

/**
 * 指定した日付の勤怠明細が存在しない時、日毎通勤回数取得APIのレスポンスの通勤往路回数・通勤復路回数に入れられる数字
 */
export const NOT_HAVE_COMMUTE_COUNT = -1 as const;

type Response = {
  isEnableStartStamp: boolean;
  isEnableEndStamp: boolean;
  isEnableRestartStamp: boolean;
  stampInDate: string;
  stampOutDate: string;
  stampReInDate: string;
  defaultAction: ClockType | null | undefined;
  commuteForwardCount: number;
  commuteBackwardCount: number;
  isPossibleFixDailyRequest: boolean;
  record: {
    recordId: string;
    fixDailyRequest: {
      requestId: string;
      status: DomainFixDailyRequest.Status;
    };
  };
};

/**
 * 日毎通勤回数取得APIのレスポンスを見て、通勤回数が取得できているかどうかを返します
 */
const existRecordWithCommuteCount = (record: Response): boolean =>
  !(
    record.commuteForwardCount === NOT_HAVE_COMMUTE_COUNT &&
    record.commuteBackwardCount === NOT_HAVE_COMMUTE_COUNT
  );

const convert = (response: Response): DomainDailyStampTime => ({
  isEnableStartStamp: response.isEnableStartStamp,
  isEnableEndStamp: response.isEnableEndStamp,
  isEnableRestartStamp: response.isEnableRestartStamp,
  stampInDate: response.stampInDate,
  stampOutDate: response.stampOutDate,
  stampReInDate: response.stampReInDate,
  defaultAction: response.defaultAction,
  commuteCount: existRecordWithCommuteCount(response)
    ? CommuteCount.convert(response)
    : null,
  isPossibleFixDailyRequest: response.isPossibleFixDailyRequest,
  record: {
    id: response.record?.recordId || '',
    fixDailyRequest: {
      id: response.record?.fixDailyRequest?.requestId || '',
      status:
        response.record?.fixDailyRequest?.status ||
        DomainFixDailyRequest.STATUS.NOT_REQUESTED,
      approver01Name: '',
      performableActionForFix:
        DomainFixDailyRequest.detectPerformableActionForFix(
          response.record?.fixDailyRequest?.status ||
            DomainFixDailyRequest.STATUS.NOT_REQUESTED
        ),
    },
  },
});

export default ((employeeId = null) =>
  Api.invoke({
    path: '/att/daily-time/get',
    param: {
      empId: employeeId,
    },
  }).then((response: Response) =>
    convert(response)
  )) as IDailyStampTimeRepository['fetch'];
