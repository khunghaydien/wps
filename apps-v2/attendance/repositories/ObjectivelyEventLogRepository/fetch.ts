import Api from '@apps/commons/api';

import {
  EventType,
  IObjectivelyEventLogRepository,
} from '@attendance/domain/models/ObjectivelyEventLog';

export type Response = {
  objectivelyEventLogList: {
    id: string;
    objectivelyEventLogSettingCode: string;
    eventType: EventType;
    eventTime: number;
    importDateTime;
  }[];
};

const fetch: IObjectivelyEventLogRepository['fetch'] = async ({
  employeeId,
  targetDate,
}) => {
  const response: Response = await Api.invoke({
    path: '/att/objectively-event-log/get',
    param: {
      empId: employeeId,
      targetDate,
    },
  });
  return response.objectivelyEventLogList.map((record) => ({
    id: record.id,
    eventType: record.eventType,
    time: record.eventTime,
    linked: record.importDateTime,
    isApplied: false,
    setting: {
      id: '',
      code: record.objectivelyEventLogSettingCode,
      name: '',
    },
  }));
};

export default fetch;
