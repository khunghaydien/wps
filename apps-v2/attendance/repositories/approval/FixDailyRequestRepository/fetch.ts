import Api from '@apps/commons/api';

import {
  AttendanceSummary,
  convert,
} from '@attendance/repositories/approval/models/AttendanceSummary';

import {
  FixDailyRequest,
  IFixDailyRequestRepository,
  Status,
} from '@attendance/domain/models/approval/FixDailyRequest';

export type Response = {
  id: string;
  status: Status;
  photoUrl: string;
  departmentName: string;
  delegatedEmployeeName: string;
  // comment: string;
  requestDate: string;
  targetDate: string;
} & AttendanceSummary;

const $convert = ({
  photoUrl: employeePhotoUrl,
  delegatedEmployeeName,
  ...request
}: Response): FixDailyRequest => {
  const convertedRequest = convert(request);
  const targetRecord = convertedRequest.records.find(
    (record) => record.recordDate === request.targetDate
  );
  return {
    id: request.id,
    requestDate: request.requestDate,
    targetDate: request.targetDate,
    submitter: {
      employee: {
        name: request.employeeName,
        code: request.employeeCode,
        photoUrl: employeePhotoUrl,
        department: {
          name: request.departmentName,
        },
      },
      delegator: {
        employee: {
          name: delegatedEmployeeName,
        },
      },
    },
    targetRecord,
    ...request,
    ...convertedRequest,
  };
};

const fetch: IFixDailyRequestRepository['fetch'] = async (requestId) => {
  const response: Response = await Api.invoke({
    path: '/att/request/fix-daily/get',
    param: {
      requestId,
    },
  });
  return $convert(response);
};

export default fetch;
