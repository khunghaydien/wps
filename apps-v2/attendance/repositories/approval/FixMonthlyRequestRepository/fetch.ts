import Api from '@apps/commons/api';

import adapter from '@apps/repositories/adapters';

import { AttendanceSummary, convert } from '../models/AttendanceSummary';
import { ApprovalHistoryList } from '@apps/domain/models/approval/request/History';
import {
  FixMonthlyRequest,
  IFixMonthlyRequestRepository,
  Status,
} from '@attendance/domain/models/approval/FixMonthlyRequest';

export type Response = {
  id: string;
  status: Status;
  employeePhotoUrl: string;
  delegatedEmployeeName: string;
  comment: string;
} & AttendanceSummary &
  ApprovalHistoryList;

const $convert = ({
  employeePhotoUrl,
  delegatedEmployeeName,
  ...response
}: Response): FixMonthlyRequest => {
  return {
    ...response,
    ...convert(response),
    submitter: {
      employee: {
        code: response.employeeCode,
        name: response.employeeName,
        photoUrl: employeePhotoUrl,
        department: {
          name: '',
        },
      },
      delegator: {
        employee: {
          name: delegatedEmployeeName,
        },
      },
    },
  };
};

const fetch: IFixMonthlyRequestRepository['fetch'] = async (id) => {
  const response = await Api.invoke({
    path: '/att/request/monthly/get',
    param: {
      requestId: id,
    },
  });
  return adapter.fromRemote(response, [$convert]);
};

export default fetch;
