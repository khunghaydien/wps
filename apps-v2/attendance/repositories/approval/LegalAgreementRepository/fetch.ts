import Api from '@apps/commons/api';

import adapter from '@apps/repositories/adapters';

import { ApprovalHistoryList } from '@apps/domain/models/approval/request/History';
import * as DomainLegalAgreementRequest from '@attendance/domain/models/approval/LegalAgreementRequest';
import { WorkSystem as LegalAgreementWorkSystem } from '@attendance/domain/models/LegalAgreementOvertime';
import {
  CODE,
  Code,
} from '@attendance/domain/models/LegalAgreementRequestType';

type Common<TCode extends Code> = {
  id: string;
  status: DomainLegalAgreementRequest.Status;
  employeeName: string;
  employeePhotoUrl: string;
  delegatedEmployeeName: string;
  typeLabel: string;
  type: TCode;
};

type MonthlyRequest = Common<typeof CODE.MONTHLY> & {
  monthlyOvertimeHours: number | null;
  monthlyOvertimeHours1MoAgo: number | null;
  monthlyOvertimeHours2MoAgo: number | null;
  specialMonthlyOvertimeHours: number | null;
  specialMonthlyOvertimeHours1MoAgo: number | null;
  specialMonthlyOvertimeHours2MoAgo: number | null;
  overtimeHoursLimit: number | null;
  extensionCount: number | null;
  extensionCountLimit: number | null;
  changedOvertimeHoursLimit: number | null;
  requestableOvertimeHoursLimit: number | null;
  reason: string | null;
  measure: string | null;
};

type YearlyRequest = Common<typeof CODE.YEARLY> & {
  yearlyOvertimeHours: number | null;
  yearlyOvertimeHours1YearAgo: number | null;
  specialYearlyOvertimeHours: number | null;
  specialYearlyOvertimeHours1YearAgo: number | null;
  overtimeHoursLimit: number | null;
  changedOvertimeHoursLimit: number | null;
  requestableOvertimeHoursLimit: number | null;
  reason: string | null;
  measure: string | null;
};

type Request = MonthlyRequest | YearlyRequest;

export type Response = {
  request: Request;
  originalRequest: Request | null;
  legalAgreementWorkSystem: LegalAgreementWorkSystem;
} & ApprovalHistoryList;

const convertSubmitter = (request: Request) => ({
  employee: {
    name: request.employeeName,
    code: '',
    photoUrl: request.employeePhotoUrl,
    department: {
      name: '',
    },
  },
  delegator: {
    employee: {
      name: request.delegatedEmployeeName,
    },
  },
});

const convertCommon = (
  request: Request
): DomainLegalAgreementRequest.BaseRequest<Code> => ({
  id: request.id,
  status: request.status,
  submitter: convertSubmitter(request),
  type: request.type,
  typeLabel: request.typeLabel,
});

const convertMonthlyRequest = (
  request: MonthlyRequest
): DomainLegalAgreementRequest.MonthlyRequest => ({
  ...(convertCommon(request) as DomainLegalAgreementRequest.BaseRequest<
    typeof CODE.MONTHLY
  >),
  monthlyOvertimeHours: request.monthlyOvertimeHours,
  monthlyOvertimeHours1MoAgo: request.monthlyOvertimeHours1MoAgo,
  monthlyOvertimeHours2MoAgo: request.monthlyOvertimeHours2MoAgo,
  specialMonthlyOvertimeHours: request.specialMonthlyOvertimeHours,
  specialMonthlyOvertimeHours1MoAgo: request.specialMonthlyOvertimeHours1MoAgo,
  specialMonthlyOvertimeHours2MoAgo: request.specialMonthlyOvertimeHours2MoAgo,
  overtimeHoursLimit: request.overtimeHoursLimit,
  extensionCount: request.extensionCount,
  extensionCountLimit: request.extensionCountLimit,
  changedOvertimeHoursLimit: request.changedOvertimeHoursLimit,
  requestableOvertimeHoursLimit: request.requestableOvertimeHoursLimit,
  reason: request.reason,
  measure: request.measure,
});

const convertYearlyRequest = (
  request: YearlyRequest
): DomainLegalAgreementRequest.YearlyRequest => ({
  ...(convertCommon(request) as DomainLegalAgreementRequest.BaseRequest<
    typeof CODE.YEARLY
  >),
  yearlyOvertimeHours: request.yearlyOvertimeHours,
  yearlyOvertimeHours1YearAgo: request.yearlyOvertimeHours1YearAgo,
  specialYearlyOvertimeHours: request.specialYearlyOvertimeHours,
  specialYearlyOvertimeHours1YearAgo:
    request.specialYearlyOvertimeHours1YearAgo,
  overtimeHoursLimit: request.overtimeHoursLimit,
  changedOvertimeHoursLimit: request.changedOvertimeHoursLimit,
  requestableOvertimeHoursLimit: request.requestableOvertimeHoursLimit,
  reason: request.reason,
  measure: request.measure,
});

const convert = (
  response: Response
): DomainLegalAgreementRequest.LegalAgreementRequest => {
  const { request, originalRequest } = response;
  switch (request.type) {
    case CODE.MONTHLY:
      return {
        request: convertMonthlyRequest(request),
        originalRequest: originalRequest
          ? convertMonthlyRequest(originalRequest as MonthlyRequest)
          : null,
        historyList: response.historyList,
        legalAgreementWorkSystem: response.legalAgreementWorkSystem,
      } as DomainLegalAgreementRequest.MonthlyLegalAgreementRequest;
    case CODE.YEARLY:
      return {
        request: convertYearlyRequest(request),
        originalRequest: originalRequest
          ? convertYearlyRequest(originalRequest as YearlyRequest)
          : null,
        historyList: response.historyList,
        legalAgreementWorkSystem: response.legalAgreementWorkSystem,
      } as DomainLegalAgreementRequest.YearlyLegalAgreementRequest;
    default:
      return null;
  }
};

const fetch: DomainLegalAgreementRequest.ILegalAgreementRequestRepository['fetch'] =
  async (id) => {
    const response = await Api.invoke({
      path: '/att/request/legal-agreement/get',
      param: {
        requestId: id,
      },
    });
    return adapter.fromRemote(response, [convert]);
  };

export default fetch;
