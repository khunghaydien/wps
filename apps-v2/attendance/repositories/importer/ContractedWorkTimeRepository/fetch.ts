import Api from '@apps/commons/api';
import { compose } from '@apps/commons/utils/FnUtil';

import { DayType } from '@apps/attendance/domain/models/AttDailyRecord';
import {
  ContractedWorkTime as DomainContractedWorkTime,
  ContractedWorkTimeRecord as DomainContractedWorkTimeRecord,
  IContractedWorkTimeRepository,
} from '@attendance/domain/models/importer/ContractedWorkTime';
import * as DomainTimeRange from '@attendance/domain/models/TimeRange';

type Record = {
  recordDate: string;
  startTime: number | null;
  endTime: number | null;
  dayType: DayType;
  rest1StartTime: number | null;
  rest1EndTime: number | null;
  rest2StartTime: number | null;
  rest2EndTime: number | null;
  rest3StartTime: number | null;
  rest3EndTime: number | null;
  rest4StartTime: number | null;
  rest4EndTime: number | null;
  rest5StartTime: number | null;
  rest5EndTime: number | null;
};

type Response = {
  summaryList: {
    startDate: string;
    endDate: string;
    workingTypeList: {
      startDate: string;
      endDate: string;
    }[];
    recordList: Record[];
  }[];
};

const mapRestTime = (record: Record): DomainTimeRange.TimeRange[] => [
  {
    startTime: record.rest1StartTime,
    endTime: record.rest1EndTime,
  },
  {
    startTime: record.rest2StartTime,
    endTime: record.rest2EndTime,
  },
  {
    startTime: record.rest3StartTime,
    endTime: record.rest3EndTime,
  },
  {
    startTime: record.rest4StartTime,
    endTime: record.rest4EndTime,
  },
  {
    startTime: record.rest5StartTime,
    endTime: record.rest5EndTime,
  },
];

const filterRestTime = (arr: DomainTimeRange.TimeRange[]) =>
  arr.filter(DomainTimeRange.hasTimes);

const createRestTimes = (record: Record): DomainTimeRange.TimeRange[] =>
  compose(
    filterRestTime,
    (arr) => arr.map(DomainTimeRange.create),
    mapRestTime
  )(record);

const convertRecord = (response: Record): DomainContractedWorkTimeRecord => ({
  recordDate: response.recordDate,
  startTime: response.startTime,
  endTime: response.endTime,
  restTimes: createRestTimes(response),
  dayType: response.dayType,
});

const convert = (
  response: Response['summaryList'][number]
): DomainContractedWorkTime => ({
  startDate: response.startDate,
  endDate: response.endDate,
  workingTypes: response.workingTypeList,
  records: new Map(
    response.recordList.map((record) => [
      record.recordDate,
      convertRecord(record),
    ])
  ),
});

export default ((param) =>
  Api.invoke({
    path: '/att/timesheet-import/default/get',
    param: {
      empId: param.employeeId,
      startDate: param.startDate,
      endDate: param.endDate,
    },
  }).then((response: Response) =>
    response?.summaryList?.map(convert)
  )) as IContractedWorkTimeRepository['fetch'];
