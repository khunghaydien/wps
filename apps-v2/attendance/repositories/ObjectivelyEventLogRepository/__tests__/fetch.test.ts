import Api from '../../../../../__tests__/mocks/ApiMock';
import fetch from '../fetch';
import { defaultValue } from './mocks/fetch.mock';

it('should return DailyObjectivelyEventLog', async () => {
  // Arrange
  const MockApi = Api.invoke as jest.Mock;
  MockApi.mockResolvedValueOnce(defaultValue);

  // Act
  const eventLogs = await fetch({
    employeeId: 'employeeId',
    targetDate: '2020-01-01',
  });

  // Assert
  expect(eventLogs).toMatchSnapshot();
  expect(MockApi).toBeCalledTimes(1);
  expect(MockApi).toBeCalledWith({
    path: '/att/objectively-event-log/get',
    param: {
      empId: 'employeeId',
      targetDate: '2020-01-01',
    },
  });
});
