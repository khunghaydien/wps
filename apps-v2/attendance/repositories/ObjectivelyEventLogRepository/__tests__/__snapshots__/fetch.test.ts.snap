// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return DailyObjectivelyEventLog 1`] = `
Array [
  Object {
    "eventType": "Entering",
    "id": "XXXXXXXXXXXXXXX",
    "isApplied": false,
    "linked": "2022-01-10T09:30:00Z",
    "setting": Object {
      "code": "CODE001",
      "id": "",
      "name": "",
    },
    "time": 570,
  },
  Object {
    "eventType": "Leaving",
    "id": "XXXXXXXXXXXXXXX",
    "isApplied": false,
    "linked": "2022-01-10T18:30:00Z",
    "setting": Object {
      "code": "CODE001",
      "id": "",
      "name": "",
    },
    "time": 1110,
  },
  Object {
    "eventType": "Entering",
    "id": "XXXXXXXXXXXXXXX",
    "isApplied": false,
    "linked": "2022-01-10T10:00:00Z",
    "setting": Object {
      "code": "CODE002",
      "id": "",
      "name": "",
    },
    "time": 600,
  },
]
`;
